pubsub/aws/snssqs/snssqs.go

```diff
@@ -251,40 +250,47 @@ func (s *snsSqs) getTopicArn(parentCtx context.Context, topic string) (string, e
 
 // get the topic ARN from the topics map. If it doesn't exist in the map, try to fetch it from AWS, if it doesn't exist
 // at all, issue a request to create the topic.
-func (s *snsSqs) getOrCreateTopic(ctx context.Context, topic string) (topicArn string, sanitizedName string, err error) {
-	s.topicsLock.Lock()
```

Comment:
Although it may _look_ like using sync.Map is as good as a solution as the previous one that used locks, it isn't. 

Per the [official docs](https://pkg.go.dev/sync#Map):

> The Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content. 

The TLDR is that sync.Map allows concurrent access to the map without issues, but it is not atomic, so in this case the code suffers from race conditions.

1. One goroutine starts and does `s.topicArns.Load`. It sees that the topic doesn't exist and starts creating one. After a delay, `s.topicsHandlers.Store` is invoked to store the topic.
2. If a separate goroutine is started at the same time, it may invoke `s.topicArns.Load` and see no topic, and then will create a new one and store it in `s.topicsHandlers.Store`.

That causes a race condition. Instead, using locks allows coordination in this case.

Reply:
ok, thanks. I'll think about maybe using xsync.MapOf (that I'm already using in the subscription manager) as I think it is lockless but safe (aka atomic). wdyt?

Reply:
No, it still wouldn't work. You need a lock that ensures coordination between multiple operations.

Imagine this were a SQL database. Using a lock is like using a transaction, so you can perform a SELECT first (reading from the map), do whatever processing you need to do (in this case, create a new connection if it doesn't exist), and then perform an INSERT to store the value.

Without a lock, it's like you don't use transactions in SQL databases. So even if there's no memory corruption because of sync.Map/xsync.MapOf, you still have the "double spending" problem.

Reply:
First I want to apologize ahead of time on everything going around this PR as the situation in Israel is very hard and my level of focus is not at its best.

As you hinted with the SQL example, the problem here is transactional isolation.
So, here, there is a need for a call to the component `Subscribe` to behave as if it is atomic and avoid contentation with other calls for `Subscribe` for the same topic. If I'm correct, in that case, all mappings read/written during subscription creation (topic --> topicArn, queue --> queueInfo, topic --> subscriptionArn & the new call to the subs mgmt `Subscribe()`) should be isolated using a barrier (and an expensive one since there are many i/o ops especially if the component provisions the aws resources). 

I think that a single lock might be too brutal as the chances for same topic/queue/subscription contentation are rare cases to be locking the component out every time subscription ops are happening and we should think about a fine-grained lock-by-topic kind of locking. wdyt?

Reply:
@ItalyPaleAle ?

Reply:
Sorry for the delay.

> First I want to apologize ahead of time on everything going around this PR as the situation in Israel is very hard and my level of focus is not at its best.

We are in no rush, and what's happening in Israel is terrible. I hope you and your loved ones are safe!

> I think that a single lock might be too brutal as the chances for same topic/queue/subscription contentation are rare cases to be locking the component out every time subscription ops are happening and we should think about a fine-grained lock-by-topic kind of locking. wdyt?

Good idea. I think this should work!